src/cart.c

```diff
@@ -0,0 +1,1814 @@
+        }
+        while ((__ed_sd_dat_rd() & 0xFF) != 0xFF);
+        addr += 512;
+    }
+    if (__ed_sd_close(__sd_type)) CART_ABORT();
+    __cart_acs_rel();
+    return 0;
+}
+
+int ed_card_wr_cart(uint32_t cart, uint32_t lba, uint32_t count)
+{
+    int i;
+    int n;
+    int resp;
+    __cart_acs_get();
+    /* SDSC takes byte address, SDHC takes LBA */
+    if (!__sd_flag) lba *= 512;
+    /* CMD25: WRITE_MULTIPLE_BLOCK */
+    if (__ed_sd_cmd(CMD25, lba) < 0) CART_ABORT();
+    if (!__sd_type)
+    {
+        /* SPI: padding (why 2 bytes?) */
+        __ed_sd_mode(ED_SD_DAT_WR, ED_SD_DAT_8b);
+        __ed_sd_dat_wr(0xFF);
+        __ed_sd_dat_wr(0xFF);
+    }
+    while (count-- > 0)
+    {
+        __ed_sd_mode(ED_SD_DAT_WR, ED_SD_DAT_8b);
+        if (!__sd_type)
+        {
+            /* SPI: data token */
+            __ed_sd_dat_wr(0xFC);
+        }
+        else
+        {
+            /* SD: start bit (why not only write F0?) */
+            __ed_sd_dat_wr(0xFF);
+            __ed_sd_dat_wr(0xF0);
+        }
+        __cart_dma_rd(__cart_buf, cart, 512);
+        /* Write data */
+        for (i = 0; i < 512; i++) __ed_sd_dat_wr(((char *)__cart_buf)[i]);
+        if (!__sd_type)
+        {
+            /* SPI: write dummy CRC */
+            for (i = 0; i < 2; i++) __ed_sd_dat_wr(0xFF);
+        }
+        else
+        {
+            /* SD: write real CRC */
+            __sd_crc16(__cart_buf, __cart_buf);
+            for (i = 0; i < 8; i++) __ed_sd_dat_wr(((char *)__cart_buf)[i]);
+            /* End bit */
+            __ed_sd_mode(ED_SD_DAT_WR, ED_SD_DAT_1b);
+            __ed_sd_dat_wr(0xFF);
+            /* Wait for start of response */
+            __ed_sd_mode(ED_SD_DAT_RD, ED_SD_DAT_1b);
+            n = 1024;
+            do
+            {
+                if (--n == 0) CART_ABORT();
+            }
+            while (__ed_sd_dat_rd() & 1);
+            /* Read response */
+            resp = 0;
+            for (i = 0; i < 3; i++) resp = resp << 1 | (__ed_sd_dat_rd() & 1);
+            if (resp != 2) CART_ABORT();
+        }
+        /* Wait for card */
+        __ed_sd_mode(ED_SD_DAT_RD, ED_SD_DAT_8b);
+        n = 65536;
+        do
+        {
+            if (--n == 0) CART_ABORT();
+        }
+        while ((__ed_sd_dat_rd() & 0xFF) != 0xFF);
+        cart += 512;
+    }
+    if (__ed_sd_close(__sd_type)) CART_ABORT();
+    __cart_acs_rel();
+    return 0;
+}
+
+int ed_card_swap(int flag)
+{
+    __cart_acs_get();
+    io_write(ED_CFG_REG, flag ?
+        (ED_CFG_SDRAM_ON|ED_CFG_SWAP) : (ED_CFG_SDRAM_ON)
+    );
+    __cart_acs_rel();
+    return 0;
+}
+
+#define SC_BASE_REG             0x1FFF0000
+#define SC_BUFFER_REG           0x1FFE0000
+
+#define SC_STATUS_REG           (SC_BASE_REG+0x00)
+#define SC_COMMAND_REG          (SC_BASE_REG+0x00)
+#define SC_DATA0_REG            (SC_BASE_REG+0x04)
+#define SC_DATA1_REG            (SC_BASE_REG+0x08)
+#define SC_VERSION_REG          (SC_BASE_REG+0x0C)
```

Comment:
This register and corresponding magic value was renamed to `IDENTIFIER` recently as it didn't carry any version information. Firmware version can be checked via `V` command. DATA0 register contains major and minor firmware version in format `[31:16]: major | [15:0]: minor` and DATA1 contains firmware revision as 32 bit number.
Major value increments when breaking changes are introduced.
Minor value increments when new commands or registers are added in non breaking fashion.
Revision value increments when no API changes were made (e.g. bug fixes)